File: runtime/erlang/src/aws_event_stream.erl
%% Shared smithy-beam Erlang Amazon Event Stream framing helpers.
-module(aws_event_stream).
-export([
    frame/2,
    decode_frames/1
]).

-type header() :: {binary(), binary() | boolean() | integer()}.
-type frame() :: #{headers := [header()], payload := binary()}.

-define(HEADER_TYPE_TRUE, 0).
-define(HEADER_TYPE_FALSE, 1).
-define(HEADER_TYPE_INTEGER, 4).
-define(HEADER_TYPE_STRING, 7).

%% @doc Encode headers and payload into a single event stream message.
-spec frame([header()], binary()) -> binary().
frame(Headers, Payload) when is_list(Headers), is_binary(Payload) ->
    HeadersBin = encode_headers(Headers),
    HeadersLen = byte_size(HeadersBin),
    PayloadLen = byte_size(Payload),
    TotalLen = HeadersLen + PayloadLen + 16,
    Prelude = <<TotalLen:32/big, HeadersLen:32/big>>,
    PreludeCrc = erlang:crc32(Prelude),
    MessageWithoutCrc =
        <<Prelude/binary, PreludeCrc:32/big, HeadersBin/binary, Payload/binary>>,
    MessageCrc = erlang:crc32(MessageWithoutCrc),
    <<MessageWithoutCrc/binary, MessageCrc:32/big>>.

%% @doc Decode one or more event stream messages from a binary body.
-spec decode_frames(binary()) -> [frame()].
decode_frames(Body) when is_binary(Body) ->
    decode_frames(Body, []).

-spec decode_frames(binary(), [frame()]) -> [frame()].
decode_frames(<<>>, Acc) ->
    lists:reverse(Acc);
decode_frames(Bin, Acc) ->
    Frame = decode_frame(Bin),
    decode_frames(maps:get(rest, Frame), [maps:without([rest], Frame) | Acc]).

-spec decode_frame(binary()) -> frame() | #{rest := binary()}.
decode_frame(Bin) ->
    case Bin of
        <<TotalLen:32/big, HeadersLen:32/big, PreludeCrc:32/big, Rest/binary>> when
            TotalLen >= 16, HeadersLen =< TotalLen - 16
        ->
            Prelude = <<TotalLen:32/big, HeadersLen:32/big>>,
            case erlang:crc32(Prelude) =:= PreludeCrc of
                false ->
                    error({bad_event_stream, invalid_prelude_crc});
                true ->
                    decode_frame_body(TotalLen, HeadersLen, PreludeCrc, Rest)
            end;
        _ ->
            error({bad_event_stream, incomplete})
    end.

-spec decode_frame_body(non_neg_integer(), non_neg_integer(), non_neg_integer(), binary()) ->
    frame() | #{rest := binary()}.
decode_frame_body(TotalLen, HeadersLen, PreludeCrc, Rest) ->
    PayloadLen = TotalLen - HeadersLen - 16,
    case Rest of
        <<HeadersBin:HeadersLen/binary, Payload:PayloadLen/binary, MessageCrc:32/big, Tail/binary>> ->
            MessageWithoutCrc =
                <<TotalLen:32/big, HeadersLen:32/big, PreludeCrc:32/big, HeadersBin/binary,
                    Payload/binary>>,
            case erlang:crc32(MessageWithoutCrc) =:= MessageCrc of
                false ->
                    error({bad_event_stream, invalid_message_crc});
                true ->
                    #{
                        headers => decode_headers(HeadersBin, []),
                        payload => Payload,
                        rest => Tail
                    }
            end;
        _ ->
            error({bad_event_stream, incomplete})
    end.

-spec encode_headers([header()]) -> binary().
encode_headers(Headers) ->
    lists:foldl(fun encode_header/2, <<>>, Headers).

-spec encode_header(header(), binary()) -> binary().
encode_header({Name, Value}, Acc) when is_binary(Name), is_binary(Value) ->
    <<Acc/binary, (byte_size(Name)):8, Name/binary, ?HEADER_TYPE_STRING:8,
        (byte_size(Value)):16/big, Value/binary>>;
encode_header({Name, true}, Acc) when is_binary(Name) ->
    <<Acc/binary, (byte_size(Name)):8, Name/binary, ?HEADER_TYPE_TRUE:8>>;
encode_header({Name, false}, Acc) when is_binary(Name) ->
    <<Acc/binary, (byte_size(Name)):8, Name/binary, ?HEADER_TYPE_FALSE:8>>;
encode_header({Name, Value}, Acc) when is_binary(Name), is_integer(Value) ->
    <<Acc/binary, (byte_size(Name)):8, Name/binary, ?HEADER_TYPE_INTEGER:8, Value:32/big>>;
encode_header({Name, _Value}, _Acc) ->
    error({bad_event_stream, {unsupported_header, Name}}).

-spec decode_headers(binary(), [header()]) -> [header()].
decode_headers(<<>>, Acc) ->
    lists:reverse(Acc);
decode_headers(Bin, Acc) ->
    case Bin of
        <<NameLen:8, Name:NameLen/binary, ?HEADER_TYPE_TRUE:8, Rest/binary>> ->
            decode_headers(Rest, [{Name, true} | Acc]);
        <<NameLen:8, Name:NameLen/binary, ?HEADER_TYPE_FALSE:8, Rest/binary>> ->
            decode_headers(Rest, [{Name, false} | Acc]);
        <<NameLen:8, Name:NameLen/binary, ?HEADER_TYPE_INTEGER:8, Value:32/big, Rest/binary>> ->
            decode_headers(Rest, [{Name, Value} | Acc]);
        <<NameLen:8, Name:NameLen/binary, ?HEADER_TYPE_STRING:8, ValueLen:16/big,
            Value:ValueLen/binary, Rest/binary>> ->
            decode_headers(Rest, [{Name, Value} | Acc]);
        _ ->
            error({bad_event_stream, invalid_header})
    end.

File: runtime/elixir/lib/aws_event_stream.ex
defmodule AwsEventStream do
  @moduledoc false

  @header_type_true 0
  @header_type_false 1
  @header_type_integer 4
  @header_type_string 7

  def frame(headers, payload) when is_list(headers) and is_binary(payload) do
    headers_bin = encode_headers(headers)
    headers_len = byte_size(headers_bin)
    payload_len = byte_size(payload)
    total_len = headers_len + payload_len + 16
    prelude = <<total_len::32-big, headers_len::32-big>>
    prelude_crc = :erlang.crc32(prelude)

    message_without_crc =
      <<prelude::binary, prelude_crc::32-big, headers_bin::binary, payload::binary>>

    message_crc = :erlang.crc32(message_without_crc)
    <<message_without_crc::binary, message_crc::32-big>>
  end

  def decode_frames(body) when is_binary(body) do
    body
    |> decode_frames([])
    |> Enum.reverse()
  end

  def encode_event_headers(event_type) do
    [
      {":event-type", event_type},
      {":message-type", "event"},
      {":content-type", "application/json"}
    ]
  end

  def header_value(headers, name) do
    Enum.find_value(headers, fn {key, value} -> if key == name, do: value end)
  end

  defp decode_frames(<<>>, acc), do: acc

  defp decode_frames(bin, acc) do
    %{rest: rest} = frame = decode_frame!(bin)
    decode_frames(rest, [Map.delete(frame, :rest) | acc])
  end

  defp decode_frame!(bin) do
    case bin do
      <<total_len::32-big, headers_len::32-big, prelude_crc::32-big, rest::binary>>
      when total_len >= 16 and headers_len <= total_len - 16 ->
        prelude = <<total_len::32-big, headers_len::32-big>>

        if :erlang.crc32(prelude) == prelude_crc do
          decode_frame_body(total_len, headers_len, prelude_crc, rest)
        else
          raise ArgumentError, "bad event stream: invalid_prelude_crc"
        end

      _ ->
        raise ArgumentError, "bad event stream: incomplete"
    end
  end

  defp decode_frame_body(total_len, headers_len, prelude_crc, rest) do
    payload_len = total_len - headers_len - 16

    case rest do
      <<headers_bin::binary-size(headers_len), payload::binary-size(payload_len),
        message_crc::32-big, tail::binary>> ->
        message_without_crc =
          <<total_len::32-big, headers_len::32-big, prelude_crc::32-big, headers_bin::binary,
            payload::binary>>

        if :erlang.crc32(message_without_crc) == message_crc do
          %{
            headers: decode_headers(headers_bin, []),
            payload: payload,
            rest: tail
          }
        else
          raise ArgumentError, "bad event stream: invalid_message_crc"
        end

      _ ->
        raise ArgumentError, "bad event stream: incomplete"
    end
  end

  defp encode_headers(headers), do: Enum.reduce(headers, <<>>, &encode_header/2)

  defp encode_header({name, value}, acc) when is_binary(name) and is_binary(value) do
    value_len = byte_size(value)
    name_len = byte_size(name)

    <<acc::binary, name_len::8, name::binary, @header_type_string::8, value_len::16-big,
      value::binary>>
  end

  defp encode_header({name, true}, acc) when is_binary(name) do
    name_len = byte_size(name)
    <<acc::binary, name_len::8, name::binary, @header_type_true::8>>
  end

  defp encode_header({name, false}, acc) when is_binary(name) do
    name_len = byte_size(name)
    <<acc::binary, name_len::8, name::binary, @header_type_false::8>>
  end

  defp encode_header({name, value}, acc) when is_binary(name) and is_integer(value) do
    name_len = byte_size(name)
    <<acc::binary, name_len::8, name::binary, @header_type_integer::8, value::32-big>>
  end

  defp encode_header({name, _value}, _acc) do
    raise ArgumentError, "bad event stream: unsupported header #{inspect(name)}"
  end

  defp decode_headers(<<>>, acc), do: Enum.reverse(acc)

  defp decode_headers(
         <<name_len::8, name::binary-size(name_len), @header_type_true::8, rest::binary>>,
         acc
       ),
       do: decode_headers(rest, [{name, true} | acc])

  defp decode_headers(
         <<name_len::8, name::binary-size(name_len), @header_type_false::8, rest::binary>>,
         acc
       ),
       do: decode_headers(rest, [{name, false} | acc])

  defp decode_headers(
         <<name_len::8, name::binary-size(name_len), @header_type_integer::8, value::32-big,
           rest::binary>>,
         acc
       ),
       do: decode_headers(rest, [{name, value} | acc])

  defp decode_headers(
         <<name_len::8, name::binary-size(name_len), @header_type_string::8, value_len::16-big,
           value::binary-size(value_len), rest::binary>>,
         acc
       ),
       do: decode_headers(rest, [{name, value} | acc])

  defp decode_headers(_bin, _acc) do
    raise ArgumentError, "bad event stream: invalid_header"
  end
end
